src/ring_theory/hahn_series.lean

```diff
@@ -0,0 +1,390 @@
+/-
+Copyright (c) 2021 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Anderson
+-/
+import order.well_founded_set
+import algebra.big_operators
+
+/-!
+# Hahn Series
+
+## Main Definitions
+  * If `Γ` is linearly ordered and `R` has zero, then `hahn_series Γ R` consists of
+  formal series over `Γ` with coefficients in `R`, whose supports are well-founded.
+  * If `R` is a (commutative) additive monoid or group, then so is `hahn_series Γ R`.
+  * If `R` is a (comm_)(semi)ring, then so is `hahn_series Γ R`.
+
+## TODO
+  * Given `[linear_ordered_add_comm_group Γ]` and `[field R]`, define `field (hahn_series Γ R)`.
+  * Build an API for the constant map `C` and the variable `X`
+  * Define Laurent series
+
+-/
+
+open_locale big_operators
+
+/-- If `Γ` is linearly ordered and `R` has zero, then `hahn_series Γ R` consists of
+  formal series over `Γ` with coefficients in `R`, whose supports are well-founded. -/
+@[ext]
+structure hahn_series (Γ : Type*) (R : Type*) [linear_order Γ] [has_zero R] :=
+(coeff : Γ → R)
+(is_wf_support : {a | coeff a ≠ 0}.is_wf)
+
+variables {Γ : Type*} {R : Type*}
+
+namespace hahn_series
+
+section zero
+variables [linear_order Γ] [has_zero R]
+
+instance : has_zero (hahn_series Γ R) :=
+⟨{ coeff := 0,
+   is_wf_support := by simp }⟩
+
+instance : inhabited (hahn_series Γ R) := ⟨0⟩
+
+@[simp]
+lemma zero_coeff {a : Γ} : (0 : hahn_series Γ R).coeff a = 0 := rfl
+
+/-- `single a r` is the Hahn series which has coefficient `r` at `a` and zero otherwise. -/
+def single (a : Γ) (r : R) : hahn_series Γ R := {
+  coeff := λ b, if a = b then r else 0,
+  is_wf_support := (set.is_wf_singleton a).mono (λ b, begin
+    simp only [and_imp, exists_prop, set.mem_singleton_iff, ite_eq_right_iff,
+      ne.def, set.mem_set_of_eq, not_forall],
+    exact λ h1 _, h1.symm,
+  end) }
+
+variables {a b : Γ} {r : R}
+
+theorem single_coeff : (single a r).coeff b = if a = b then r else 0 := rfl
+
+@[simp]
+theorem single_coeff_same : (single a r).coeff a = r := if_pos rfl
+
+@[simp]
+theorem single_coeff_of_ne (h : a ≠ b) : (single a r).coeff b = 0 := if_neg h
+
+@[simp]
+lemma single_support_of_ne (h : r ≠ 0) : {b | (single a r).coeff b ≠ 0} = {a} :=
+begin
+  ext b,
+  simp only [set.mem_singleton_iff, ne.def, set.mem_set_of_eq],
+  split,
+  { contrapose!,
+    exact λ h', single_coeff_of_ne (ne.symm h') },
+  { rintro rfl,
+    simp [h] }
+end
+
+@[simp]
+lemma single_eq_zero : (single a (0 : R)) = 0 := by { ext, simp [single_coeff] }
```

Comment:
This proof becomes `ext (pi.single_zero _ _)` or similar with my suggestion to use `pi.single`